File: bridges/mcp_bridge.py
"""Bridge to any MCP (Model Context Protocol) server.

Supports two transports:
  - stdio: launch a local command and speak JSON-RPC over its stdin/stdout
  - http:  POST JSON-RPC to a URL

We implement just enough of MCP to call `initialize`, `tools/list`, and
`tools/call`. No external `mcp` SDK dependency — keeps Atlas slim.
"""
from __future__ import annotations

import asyncio
import json
import shlex
from typing import List

import httpx

from .base import Bridge, BridgeError
from .spec import ToolSpec, BridgeSpec, Capability


PROTOCOL_VERSION = "2024-11-05"


class _StdioMCP:
    """JSON-RPC over a child process's stdio. One outstanding request map."""

    def __init__(self, command: str, env: dict | None = None, cwd: str | None = None):
        self._command = command
        self._env = env
        self._cwd = cwd
        self._proc: asyncio.subprocess.Process | None = None
        self._reader_task: asyncio.Task | None = None
        self._next_id = 1
        self._pending: dict[int, asyncio.Future] = {}
        self._lock = asyncio.Lock()

    async def start(self) -> None:
        if self._proc is not None:
            return
        argv = shlex.split(self._command)
        self._proc = await asyncio.create_subprocess_exec(
            *argv,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=self._env,
            cwd=self._cwd,
        )
        self._reader_task = asyncio.create_task(self._reader())

    async def _reader(self) -> None:
        assert self._proc is not None and self._proc.stdout is not None
        try:
            while True:
                line = await self._proc.stdout.readline()
                if not line:
                    break
                try:
                    msg = json.loads(line.decode().strip())
                except Exception:
                    continue
                if "id" in msg and msg["id"] in self._pending:
                    fut = self._pending.pop(msg["id"])
                    if not fut.done():
                        fut.set_result(msg)
        except asyncio.CancelledError:
            pass

    async def call(self, method: str, params: dict | None = None, timeout: float = 60) -> dict:
        if self._proc is None or self._proc.stdin is None:
            raise BridgeError("mcp stdio: process not started")
        async with self._lock:
            req_id = self._next_id
            self._next_id += 1
        fut: asyncio.Future = asyncio.get_event_loop().create_future()
        self._pending[req_id] = fut
        payload = {"jsonrpc": "2.0", "id": req_id, "method": method, "params": params or {}}
        self._proc.stdin.write((json.dumps(payload) + "\n").encode())
        await self._proc.stdin.drain()
        try:
            msg = await asyncio.wait_for(fut, timeout=timeout)
        except asyncio.TimeoutError:
            self._pending.pop(req_id, None)
            raise BridgeError(f"mcp call '{method}' timed out after {timeout}s")
        if "error" in msg:
            err = msg["error"]
            raise BridgeError(f"mcp error {err.get('code')}: {err.get('message')}")
        return msg.get("result", {})

    async def notify(self, method: str, params: dict | None = None) -> None:
        if self._proc is None or self._proc.stdin is None:
            return
        payload = {"jsonrpc": "2.0", "method": method, "params": params or {}}
        self._proc.stdin.write((json.dumps(payload) + "\n").encode())
        await self._proc.stdin.drain()

    async def close(self) -> None:
        if self._reader_task:
            self._reader_task.cancel()
            try:
                await self._reader_task
            except (asyncio.CancelledError, Exception):
                pass
        if self._proc:
            try:
                self._proc.terminate()
                await asyncio.wait_for(self._proc.wait(), timeout=5)
            except Exception:
                try:
                    self._proc.kill()
                except Exception:
                    pass


class _HTTPMCP:
    def __init__(self, url: str, headers: dict | None = None):
        self._url = url
        self._headers = headers or {}
        self._client: httpx.AsyncClient | None = None
        self._next_id = 1

    async def start(self) -> None:
        if self._client is None:
            self._client = httpx.AsyncClient(headers=self._headers, timeout=60)

    async def call(self, method: str, params: dict | None = None, timeout: float = 60) -> dict:
        if self._client is None:
            await self.start()
        assert self._client is not None
        req_id = self._next_id
        self._next_id += 1
        payload = {"jsonrpc": "2.0", "id": req_id, "method": method, "params": params or {}}
        resp = await self._client.post(self._url, json=payload, timeout=timeout)
        resp.raise_for_status()
        msg = resp.json()
        if "error" in msg:
            err = msg["error"]
            raise BridgeError(f"mcp error {err.get('code')}: {err.get('message')}")
        return msg.get("result", {})

    async def notify(self, method: str, params: dict | None = None) -> None:
        if self._client is None:
            return
        payload = {"jsonrpc": "2.0", "method": method, "params": params or {}}
        await self._client.post(self._url, json=payload)

    async def close(self) -> None:
        if self._client:
            await self._client.aclose()


class MCPBridge(Bridge):
    """`config`:
        transport: "stdio" | "http"        (required)
        # stdio
        command:   str                     — full command line (e.g. "uvx mcp-server-fetch")
        env:       dict[str,str]
        cwd:       str
        # http
        url:       str
        headers:   dict[str,str]
    """

    kind = "mcp"

    def __init__(self, spec: BridgeSpec):
        super().__init__(spec)
        self._client: _StdioMCP | _HTTPMCP | None = None
        self._tools_cache: list[ToolSpec] = []

    async def connect(self) -> None:
        if self._client is not None:
            return
        cfg = self.spec.config
        transport = cfg.get("transport", "stdio")
        if transport == "stdio":
            cmd = cfg.get("command")
            if not cmd:
                raise BridgeError("mcp stdio: 'command' is required")
            self._client = _StdioMCP(cmd, env=cfg.get("env"), cwd=cfg.get("cwd"))
        elif transport == "http":
            url = cfg.get("url")
            if not url:
                raise BridgeError("mcp http: 'url' is required")
            self._client = _HTTPMCP(url, headers=cfg.get("headers", {}))
        else:
            raise BridgeError(f"mcp: unsupported transport '{transport}'")
        await self._client.start()

        # Initialize handshake
        try:
            await self._client.call("initialize", {
                "protocolVersion": PROTOCOL_VERSION,
                "capabilities": {},
                "clientInfo": {"name": "Atlas", "version": "0.1"},
            }, timeout=15)
            await self._client.notify("notifications/initialized")
        except Exception as e:
            raise BridgeError(f"mcp initialize failed: {e}") from e

    async def discover(self) -> List[ToolSpec]:
        await self.connect()
        assert self._client is not None
        try:
            result = await self._client.call("tools/list")
        except Exception as e:
            raise BridgeError(f"mcp tools/list failed: {e}") from e
        tools_raw = result.get("tools", [])
        out: list[ToolSpec] = []
        for t in tools_raw:
            name = t.get("name")
            if not name:
                continue
            out.append(ToolSpec(
                name=name,
                description=(t.get("description") or "")[:300],
                parameters=t.get("inputSchema") or {"type": "object", "properties": {}, "required": []},
                handler={"op": "call_tool", "tool_name": name},
            ))
        self._tools_cache = out
        return out

    async def invoke(self, tool: ToolSpec, arguments: dict) -> str:
        if self._client is None:
            await self.connect()
        assert self._client is not None
        if tool.handler.get("op") != "call_tool":
            raise BridgeError(f"mcp: unknown op {tool.handler!r}")
        upstream_name = tool.handler["tool_name"]
        try:
            result = await self._client.call("tools/call", {
                "name": upstream_name,
                "arguments": arguments,
            })
        except Exception as e:
            raise BridgeError(f"mcp call '{upstream_name}' failed: {e}") from e
        # MCP tool results are a list of content parts.
        parts = result.get("content") or []
        chunks: list[str] = []
        for p in parts:
            if p.get("type") == "text" and isinstance(p.get("text"), str):
                chunks.append(p["text"])
            else:
                chunks.append(json.dumps(p))
        if result.get("isError"):
            return "[mcp tool returned error]\n" + "\n".join(chunks)
        return "\n".join(chunks) if chunks else "(empty)"

    async def aclose(self) -> None:
        if self._client is not None:
            await self._client.close()
            self._client = None

    @staticmethod
    def default_capabilities() -> list[str]:
        return [Capability.CONTROL.value, Capability.TELEMETRY.value]

File: cli-ink/src/bridges/mcp-bridge.ts
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { request } from "undici";
import { Bridge } from "./base.js";
import { BridgeError, type ToolSpec } from "./types.js";

const PROTOCOL_VERSION = "2024-11-05";

interface JsonRpcResponse {
  jsonrpc: "2.0";
  id?: number;
  result?: any;
  error?: { code: number; message: string };
}

function parseCommand(cmd: string): { argv: string[] } {
  // Minimal shell-style splitter — handles quoted args. Good enough for the
  // command lines registries hand us ("npx -y @x/y --arg 'with space'").
  const out: string[] = [];
  let cur = "";
  let quote: '"' | "'" | "" = "";
  for (let i = 0; i < cmd.length; i++) {
    const c = cmd[i]!;
    if (quote) {
      if (c === quote) quote = "";
      else cur += c;
    } else if (c === '"' || c === "'") {
      quote = c;
    } else if (c === " " || c === "\t") {
      if (cur) {
        out.push(cur);
        cur = "";
      }
    } else {
      cur += c;
    }
  }
  if (cur) out.push(cur);
  return { argv: out };
}

class StdioMCP {
  private proc?: ChildProcessWithoutNullStreams;
  private nextId = 1;
  private pending = new Map<number, (msg: JsonRpcResponse) => void>();
  private buf = "";

  constructor(
    private command: string,
    private env?: Record<string, string>,
    private cwd?: string,
  ) {}

  async start(): Promise<void> {
    if (this.proc) return;
    const { argv } = parseCommand(this.command);
    if (argv.length === 0) throw new BridgeError("mcp stdio: empty command");
    const [cmd, ...rest] = argv;
    this.proc = spawn(cmd!, rest, {
      stdio: ["pipe", "pipe", "pipe"],
      env: this.env ? { ...process.env, ...this.env } : process.env,
      cwd: this.cwd,
    }) as ChildProcessWithoutNullStreams;

    this.proc.stdout.setEncoding("utf8");
    this.proc.stdout.on("data", (chunk: string) => this.onStdout(chunk));
    this.proc.on("error", (err) => {
      // Reject any pending calls; future calls will fail at write-time.
      for (const [, resolve] of this.pending) {
        resolve({ jsonrpc: "2.0", error: { code: -32000, message: String(err.message) } });
      }
      this.pending.clear();
    });
    this.proc.on("exit", (code) => {
      for (const [, resolve] of this.pending) {
        resolve({
          jsonrpc: "2.0",
          error: { code: -32001, message: `mcp server exited (code ${code})` },
        });
      }
      this.pending.clear();
    });
  }

  private onStdout(chunk: string): void {
    this.buf += chunk;
    let nl: number;
    while ((nl = this.buf.indexOf("\n")) >= 0) {
      const line = this.buf.slice(0, nl).trim();
      this.buf = this.buf.slice(nl + 1);
      if (!line) continue;
      let msg: JsonRpcResponse;
      try {
        msg = JSON.parse(line);
      } catch {
        continue;  // ignore malformed lines (some servers print debug to stdout)
      }
      if (typeof msg.id === "number" && this.pending.has(msg.id)) {
        const resolve = this.pending.get(msg.id)!;
        this.pending.delete(msg.id);
        resolve(msg);
      }
    }
  }

  async call(method: string, params?: any, timeoutMs = 60_000): Promise<any> {
    if (!this.proc?.stdin) throw new BridgeError("mcp stdio: process not started");
    const id = this.nextId++;
    const payload = JSON.stringify({ jsonrpc: "2.0", id, method, params: params ?? {} }) + "\n";
    const result: JsonRpcResponse = await new Promise((resolve, reject) => {
      this.pending.set(id, resolve);
      const t = setTimeout(() => {
        this.pending.delete(id);
        reject(new BridgeError(`mcp call '${method}' timed out after ${timeoutMs}ms`));
      }, timeoutMs);
      this.proc!.stdin.write(payload, (err) => {
        if (err) {
          clearTimeout(t);
          this.pending.delete(id);
          reject(new BridgeError(`mcp stdin write failed: ${err.message}`));
        }
      });
      // Resolve also clears the timer once it fires.
      const wrapped = this.pending.get(id)!;
      this.pending.set(id, (msg) => {
        clearTimeout(t);
        wrapped(msg);
      });
    });
    if (result.error) throw new BridgeError(`mcp error ${result.error.code}: ${result.error.message}`);
    return result.result;
  }

  notify(method: string, params?: any): void {
    if (!this.proc?.stdin) return;
    const payload = JSON.stringify({ jsonrpc: "2.0", method, params: params ?? {} }) + "\n";
    this.proc.stdin.write(payload);
  }

  async close(): Promise<void> {
    if (!this.proc) return;
    try {
      this.proc.kill();
    } catch {
      /* ignore */
    }
    this.proc = undefined;
  }
}

class HttpMCP {
  private nextId = 1;
  constructor(
    private url: string,
    private headers: Record<string, string> = {},
  ) {}

  async start(): Promise<void> {
    /* nothing — request-per-call */
  }

  async call(method: string, params?: any, timeoutMs = 60_000): Promise<any> {
    const id = this.nextId++;
    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), timeoutMs);
    try {
      const res = await request(this.url, {
        method: "POST",
        headers: { "content-type": "application/json", ...this.headers },
        body: JSON.stringify({ jsonrpc: "2.0", id, method, params: params ?? {} }),
        signal: controller.signal,
      });
      const text = await res.body.text();
      const msg = JSON.parse(text) as JsonRpcResponse;
      if (msg.error) throw new BridgeError(`mcp error ${msg.error.code}: ${msg.error.message}`);
      return msg.result;
    } finally {
      clearTimeout(t);
    }
  }

  notify(_method: string, _params?: any): void {
    // No-op for HTTP MCP — initialization notifications don't need a response.
  }

  async close(): Promise<void> {
    /* nothing */
  }
}

export class MCPBridge extends Bridge {
  static override kind = "mcp";
  private client?: StdioMCP | HttpMCP;

  override async connect(): Promise<void> {
    if (this.client) return;
    const cfg = this.spec.config;
    const transport = String(cfg.transport ?? "stdio");
    if (transport === "stdio") {
      if (!cfg.command) throw new BridgeError("mcp stdio: 'command' is required");
      this.client = new StdioMCP(cfg.command, cfg.env, cfg.cwd);
    } else if (transport === "http") {
      if (!cfg.url) throw new BridgeError("mcp http: 'url' is required");
      this.client = new HttpMCP(cfg.url, cfg.headers ?? {});
    } else {
      throw new BridgeError(`mcp: unsupported transport '${transport}'`);
    }
    await this.client.start();
    try {
      await this.client.call("initialize", {
        protocolVersion: PROTOCOL_VERSION,
        capabilities: {},
        clientInfo: { name: "Atlas", version: "0.1" },
      }, 15_000);
      this.client.notify("notifications/initialized");
    } catch (e: any) {
      throw new BridgeError(`mcp initialize failed: ${e?.message ?? e}`);
    }
  }

  override async discover(): Promise<ToolSpec[]> {
    await this.connect();
    const result = await this.client!.call("tools/list");
    const upstream = (result?.tools ?? []) as Array<{
      name?: string;
      description?: string;
      inputSchema?: any;
    }>;
    const out: ToolSpec[] = [];
    for (const t of upstream) {
      if (!t.name) continue;
      out.push({
        name: t.name,
        description: (t.description ?? "").slice(0, 300),
        parameters: t.inputSchema ?? { type: "object", properties: {}, required: [] },
        handler: { op: "call_tool", tool_name: t.name },
      });
    }
    return out;
  }

  override async invoke(tool: ToolSpec, args: Record<string, any>): Promise<string> {
    if (!this.client) await this.connect();
    if (tool.handler?.op !== "call_tool") {
      throw new BridgeError(`mcp: unknown op ${JSON.stringify(tool.handler)}`);
    }
    const upstreamName = String(tool.handler.tool_name);
    const result = await this.client!.call("tools/call", { name: upstreamName, arguments: args ?? {} });
    const parts = (result?.content ?? []) as any[];
    const chunks: string[] = [];
    for (const p of parts) {
      if (p?.type === "text" && typeof p.text === "string") chunks.push(p.text);
      else chunks.push(JSON.stringify(p));
    }
    if (result?.isError) return "[mcp tool returned error]\n" + chunks.join("\n");
    return chunks.length ? chunks.join("\n") : "(empty)";
  }

  override async aclose(): Promise<void> {
    if (this.client) {
      await this.client.close();
      this.client = undefined;
    }
  }

  static defaultCapabilities(): string[] {
    return ["control", "telemetry"];
  }
}
